services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/multipart/MultipartUploadHelper.java

```diff
@@ -78,9 +79,19 @@ CompletableFuture<CreateMultipartUploadResponse> createMultipartUpload(PutObject
     CompletableFuture<CompleteMultipartUploadResponse> completeMultipartUpload(CompletableFuture<PutObjectResponse> returnFuture,
                                  String uploadId,
                                  CompletedPart[] completedParts,
-                                 PutObjectRequest putObjectRequest) {
+                                 PutObjectRequest putObjectRequest,
+                                 PublisherListener<Long> progressListener) {
         CompletableFuture<CompleteMultipartUploadResponse> future =
-            genericMultipartHelper.completeMultipartUpload(putObjectRequest, uploadId, completedParts);
+            genericMultipartHelper.completeMultipartUpload(putObjectRequest, uploadId, completedParts)
+                                  .whenComplete((r, t) -> {
+                                      if (t != null) {
+                                          // leave exception handling to the subscriber which checks failureActionInitiated flag
+                                      } else {
+                                          if (progressListener != null) {
+                                              progressListener.subscriberOnComplete();
```

Comment:
Can this be done in `genericMultipartHelper.handleExceptionOrResponse(putObjectRequest, returnFuture, uploadId)`?